src/openrct2-ui/windows/NewCampaign.cpp

```diff
@@ -50,376 +49,354 @@ static rct_widget window_new_campaign_widgets[] = {
+            if (rideEntry != nullptr)
             {
-                if (itemType != ShopItem::None && GetShopItemDescriptor(itemType).IsFoodOrDrink())
+                for (const auto itemType : rideEntry->shop_item)
                 {
-                    items |= EnumToFlag(itemType);
+                    if (itemType != ShopItem::None && GetShopItemDescriptor(itemType).IsFoodOrDrink())
+                    {
+                        items |= EnumToFlag(itemType);
+                    }
                 }
             }
         }
-    }
 
-    //
-    auto numItems = 0;
-    for (auto i = 0; i < 64; i++)
-    {
-        if (items & (1ULL << i))
+        //
+        auto numItems = 0;
+        for (auto i = 0; i < 64; i++)
         {
-            window_new_campaign_shop_items[numItems++] = i;
+            if (items & (1ULL << i))
+            {
+                ShopItems[numItems++] = i;
+            }
         }
+        ShopItems[numItems] = 255;
     }
-    window_new_campaign_shop_items[numItems] = 255;
-}
 
-/**
- *
- *  rct2: 0x0069E50B
- */
-static void window_new_campaign_mouseup(rct_window* w, rct_widgetindex widgetIndex)
-{
-    switch (widgetIndex)
+public:
+    void RefreshRides()
     {
-        case WIDX_CLOSE:
-            window_close(w);
-            break;
-        case WIDX_START_BUTTON:
+        // Get all applicable rides
+        RideList.clear();
+        for (const auto& curRide : GetRideManager())
         {
-            auto gameAction = ParkMarketingAction(w->campaign.campaign_type, w->campaign.RideId, w->campaign.no_weeks);
-            gameAction.SetCallback([](const GameAction* ga, const GameActions::Result* result) {
-                if (result->Error == GameActions::Status::Ok)
+            if (curRide.status == RIDE_STATUS_OPEN)
+            {
+                if (!curRide.GetRideTypeDescriptor().HasFlag(
+                        RIDE_TYPE_FLAG_IS_SHOP | RIDE_TYPE_FLAG_SELLS_FOOD | RIDE_TYPE_FLAG_SELLS_DRINKS
+                        | RIDE_TYPE_FLAG_IS_TOILET))
                 {
-                    window_close_by_class(WC_NEW_CAMPAIGN);
+                    RideList.push_back(curRide.id);
                 }
-            });
-            GameActions::Execute(&gameAction);
-            break;
+            }
+        }
+
+        // Take top 128 most valuable rides
+        if (RideList.size() > Dropdown::ItemsMaxSize)
+        {
+            qsort(RideList.data(), RideList.size(), sizeof(ride_id_t), RideValueCompare);
+            RideList.resize(Dropdown::ItemsMaxSize);
         }
+
+        // Sort rides by name
+        qsort(RideList.data(), RideList.size(), sizeof(ride_id_t), RideNameCompare);
     }
-}
 
-/**
- *
- *  rct2: 0x0069E51C
- */
-static void window_new_campaign_mousedown(rct_window* w, rct_widgetindex widgetIndex, rct_widget* widget)
-{
-    rct_widget* dropdownWidget;
+    void OnOpen() override
+    {
+        widgets = window_new_campaign_widgets;
```

Comment:
Replace all uses of `window_new_campaign_widgets` with `w->widgets`, with the exception of the assignment of `w->widgets`